### PR TITLE
fix: skip MarshalXML for filter if empty

### LINKED ISSUE #10847: GetBucketLifecycleConfiguration API returns empty <Filter> and <Prefix> tag
According to the description of the document https://docs.min.io/docs/java-client-api-reference.html#getBucketLifecycle, calling the getBucketLifecycle method is directly abnormal
```
---------START-HTTP---------
GET /test?lifecycle= HTTP/1.1
Host: 52.82.91.219:9000
Accept-Encoding: identity
User-Agent: MinIO (Windows Server 2019; amd64) minio-java/8.0.0
Content-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==
x-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
x-amz-date: 20201106T065912Z
Authorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20201106/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*

HTTP/1.1 200
Accept-Ranges: bytes
Content-Length: 299
Content-Security-Policy: block-all-mixed-content
Content-Type: application/xml
Server: MinIO/RELEASE.2020-10-28T08-16-50Z
Vary: Origin
X-Amz-Request-Id: 1644D8FDDF713B9C
X-Xss-Protection: 1; mode=block
Date: Fri, 06 Nov 2020 06:59:12 GMT

----------END-HTTP----------
```
response
``` xml
<LifecycleConfiguration>
<Rule>
<ID>b4f66e8f-16db-4476-b012-f3d751d54e50</ID>
<Status>Enabled</Status>
<Filter><Prefix></Prefix></Filter>
<Expiration><Days>91</Days></Expiration>
<NoncurrentVersionExpiration><NoncurrentDays>91</NoncurrentDays></NoncurrentVersionExpiration>
</Rule>
</LifecycleConfiguration>
```

exception:
org.simpleframework.xml.core.PersistenceException: Constructor not matched for class io.minio.messages.RuleFilter
image

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
fix: skip MarshalXML for filter if empty

## Motivation and Context
fixes #10847

## How to test this PR?
As per the reproduer in #10847 

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed

### CODE REVIEW

File: pkg/bucket/lifecycle/filter.go [hunk 1]
```diff
@@ -38,6 +38,11 @@ type Filter struct {
 // MarshalXML - produces the xml representation of the Filter struct
 // only one of Prefix, And and Tag should be present in the output.
 func (f Filter) MarshalXML(e *xml.Encoder, start xml.StartElement) error {
+	if f.And.isEmpty() && f.Tag.IsEmpty() && f.Prefix == "" {
```
Comment (balamurugana) [inline pkg/bucket/lifecycle/filter.go:41]: As setting empty prefix value is valid, we skip sending empty prefix as per this fix.

XML used in `SetBucketLifecycleConfiguration` API is `<Filter><Prefix></Prefix></Filter>`, with this fix, we don't get back that.

Comment (harshavardhana) [reply pkg/bucket/lifecycle/filter.go:41]: AWS doesn't send it back either @balamurugana - so it is simply filtered out if its not set.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#b6dfde3f-e127-4929-a5a6-ed8f43bc639f) |
<a name="b6dfde3f-e127-4929-a5a6-ed8f43bc639f"></a>
### 10851-4c63264/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:30293
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 0070a942bcb8:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 9 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 1 minutes and 29 seconds
(4/15) Running aws-sdk-ruby tests ... done in 20 seconds
(5/15) Running awscli tests ... done in 2 minutes and 55 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... FAILED in 55 seconds
{
  "name": "mc",
  "duration": "23159",
  "function": "test_get_object_multipart",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-31508 --quiet --no-color cp myminio/mc-test-bucket-11382/mc-test-object-29020 mc-test-object-29020.downloaded  >>> \n`myminio/mc-test-bucket-11382/mc-test-object-29020` -> `mc-test-object-29020.downloaded`\nmc: <ERROR> Failed to copy `http://minio-dev6.minio.io:30293/mc-test-bucket-11382/mc-test-object-29020`. read tcp 172.17.0.8:54076->72.28.97.55:30293: read: connection reset by peer\nTotal: 0 B, Transferred: 60.78 MiB, Speed: 35.73 MiB/s"
}
(7/15) Running minio-dotnet tests ... done in 1 minutes and 45 seconds
(8/15) Running minio-go tests ... FAILED in 1 minutes and 57 seconds
{
  "args": {
    "bucketName": "minio-go-test-2d2ct44opsia609h",
    "objectName": "test-object",
    "opts": "",
    "size": -1
  },
  "duration": 1279,
  "function": "PutObject(bucketName, objectName, reader,size,opts)",
  "message": "Unexpected size",
  "name": "minio-go: testPutObjectStreaming",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... done in 10 minutes and 28 seconds
(9/15) Running minio-js tests ... FAILED in 46 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 80,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:682:10) at endReadableNT (_stream_readable.js:1252:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(9/15) Running minio-py tests ... done in 19 minutes and 35 seconds
(10/15) Running s3cmd tests ... done in 2 minutes and 14 seconds
(11/15) Running s3select tests ... done in 38 seconds
(12/15) Running security tests ... done in 0 seconds

Executed 12 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: pkg/bucket/lifecycle/filter.go (hunk 1 above)
Comment (harshavardhana) [reply pkg/bucket/lifecycle/filter.go:41]: Ah you are right @balamurugana looks like MinIO server is sending correct values. We don't need this PR.

**Pull request closed without merging**